File: 2022/day4/solve.py
data = [line.strip().split(',') for line in open('input.txt', 'r')]

def checkOverlap(first, second, partB = False):
    def check(a, b):
        return (a[0] >= b[0] and a[0 if partB else 1] <= b[1])
    return check(first, second) or check(second, first)

assert(not checkOverlap([2,4],[6,8]))
assert(checkOverlap([6,6],[4,6]))
assert(checkOverlap([4,6],[6,6]))
assert(checkOverlap([1,1],[1,1]))
assert(checkOverlap([1,5],[2,6], True))
assert(checkOverlap([2,6],[1,5], True))

overlapTimes = 0
overlapAtAll = 0
for line in data:
    first = [int(e) for e in line[0].split('-')]
    second = [int(e) for e in line[1].split('-')]
    if checkOverlap(first, second):
        overlapTimes += 1
    if checkOverlap(first, second, True):
        overlapAtAll += 1

print(overlapTimes)
print(overlapAtAll)

File: 2023/day1/solve.go
package main

import (
	"saneri/aoc/utils"
	"fmt"
	"log"
	"regexp"
	"strconv"
	"strings"
)

func stringToNumber(str string) int {
	wordToNumber := map[string]int{
		"one": 1,
		"two": 2,
		"three": 3,
		"four": 4,
		"five": 5,
		"six": 6,
		"seven": 7,
		"eight": 8,
		"nine": 9,
	}

	if (len(str) > 1) {
		number := wordToNumber[str]
		if (number == 0){
			log.Fatal("fatal: ", str)
		}
		return number
	}
	number, err := strconv.Atoi(str)
	if (err != nil) {
		log.Fatal(err)
	}
	return number
}

func main() {
	data := utils.ReadInput("input.txt")
	sum := 0
	re := regexp.MustCompile("[1-9]|(one)|(two)|(three)|(four)|(five)|(six)|(seven)|(eight)|(nine)")
	for _, line := range data {
		// had to create this monster for loop because I found no ways of finding overlapping matches in go regex
		digits := []string{}
		for (len(line) > 0) {
			str := re.FindString(line)
			if (len(strings.TrimSpace(str)) > 0) {
				digits = append(digits, str)
			}
			line = line[1:]
		}
		first := stringToNumber(digits[0])
		second := stringToNumber(digits[len(digits)-1])
		sum += first * 10 + second
	}
	fmt.Println(sum)
}